### PR TITLE
Feat/aut722 Rework for a more modular GUI / Settings

### PR DESCRIPTION
**Pull request by bertrandboudaud**
Rework for a more modular GUI / Settings.

- Removed a lot of redundant code (Session handler, GUI).
- Created some dedicated widgets from the widget we were creating from within GUI.
- Reduced the amount of code to write to add a new widget to the GUI:
  - More modular. a list of top, bottom, and left windows are created, the GUI is responsible for calling the draw methods on it.
  - Adding a widget is to declare it on the global list of widgets that will be used, set added on of the layout lists (top, bottom, left), and add a menu to activate it.
  - Added visible flag to widget class to avoid creating a variable for each created widget
  - Added title (not same as id..) to the definition of a widget

### CODE REVIEW

File: src/examples/source/SmartPeakGUI.cpp [hunk 1]
```diff
@@ -295,180 +335,82 @@ int main(int argc, char** argv)
...
+      quickInfoText_->clearErrorMessages();
+      if (exceeding_plot_points_) quickInfoText_->addErrorMessage("Plot rendering limit reached.  Not plotting all selected data.");
+      if (exceeding_table_size_) quickInfoText_->addErrorMessage("Table rendering limit reached.  Not showing all selected data.");
+      if (ran_integrity_check_ && integrity_check_failed_) quickInfoText_->addErrorMessage("Integrity check failed.  Check the `Information` log.");
+      if (ran_integrity_check_ && !integrity_check_failed_) quickInfoText_->addErrorMessage("Integrity check passed.");
 
       // Session cache updates from e.g., single file imports
       if (file_loading_is_done_ && file_picker_.fileWasLoaded() && update_session_cache_) {
         if (file_picker_.getProcessorName() == "LoadSequenceParameters") {
-          session_handler_.sequence_table_headers.resize(0);
```
Comment (dmccloskey) [inline src/examples/source/SmartPeakGUI.cpp:319]: +1 to `.clear()`

File: src/examples/source/SmartPeakGUI.cpp [hunk 2]
```diff
@@ -295,180 +335,82 @@ int main(int argc, char** argv)
...
-          session_handler_.comp_group_qcs_table_headers.resize(0);
-          session_handler_.comp_group_qcs_table_body.resize(0, 0);
+          comp_group_qc_main_window_->table_data_.clear();
         }
         update_session_cache_ = false;
       }
 
-    // Pop-ups
-    if (popup_file_picker_)
+    // ======================================
```
Comment (dmccloskey) [inline src/examples/source/SmartPeakGUI.cpp:384]: Perhaps a small detail, but this is quite helpful in increasing readability.

File: src/examples/source/SmartPeakGUI.cpp [hunk 3]
```diff
@@ -747,80 +690,199 @@ int main(int argc, char** argv)
...
-      || show_comp_background_filters_table || show_comp_group_background_filters_table || show_comp_background_qcs_table || show_comp_group_background_qcs_table
-      || show_comp_rsd_estimations_table || show_comp_group_rsd_estimations_table || show_comp_background_estimations_table || show_comp_group_background_estimations_table
-      || show_feature_table || show_feature_pivot_table
-      || show_chromatogram_line_plot || show_spectra_line_plot || show_feature_line_plot || show_feature_heatmap_plot || show_calibrators_line_plot;
-    show_bottom_window_ = show_info_ || show_log_;
-    show_left_window_ = show_injection_explorer || show_transitions_explorer || show_features_explorer || show_spectrum_explorer;
+    // ======================================
+    // Window size computation
+    // ======================================
+    bool show_top_window_ = std::find_if(top_windows.begin(), top_windows.end(), [](const auto& w) { return w->visible_; }) != top_windows.end();
```
Comment (dmccloskey) [inline src/examples/source/SmartPeakGUI.cpp:693]: Very elegant.

File: src/examples/source/SmartPeakGUI.cpp [hunk 4]
```diff
@@ -747,80 +690,199 @@ int main(int argc, char** argv)
...
-        ImGuiWindowFlags_NoTitleBar |
-        ImGuiWindowFlags_NoMove |
-        ImGuiWindowFlags_NoCollapse |
-        ImGuiWindowFlags_NoFocusOnAppearing;
-      ImGui::Begin("Left window", NULL, left_window_flags);
+      ImGui::Begin("Left window", NULL, window_flags);
       if (ImGui::BeginTabBar("Left window tab bar", ImGuiTabBarFlags_Reorderable))
       {
-        if (show_injection_explorer && ImGui::BeginTabItem("Injections", &show_injection_explorer))
+        for (auto& widget : left_windows)
```
Comment (dmccloskey) [inline src/examples/source/SmartPeakGUI.cpp:871]: Wrapping all of the tabs into a `std::vector` was quite the design choice for reducing code and redundancy!

File: src/smartpeak/include/SmartPeak/core/SessionHandler.h [hunk 5]
```diff
@@ -217,96 +230,23 @@ namespace SmartPeak
     int getNSelectedFeatureMetaValuesPlot();
 
     // data for the injection explorer
-    Eigen::Tensor<std::string, 1> injection_explorer_checkbox_headers;
-    Eigen::Tensor<bool, 2> injection_explorer_checkbox_body;
-    Eigen::Tensor<bool, 1> injection_explorer_checked_rows;
+    ExplorerData injection_explorer_data;
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/core/SessionHandler.h:233]: Wrapping these into their own structs was a nice choice.

File: src/smartpeak/include/SmartPeak/ui/SpectraPlotWidget.h [hunk 6]
```diff
@@ -17,8 +17,9 @@ namespace SmartPeak
   public:
     SpectraPlotWidget(SessionHandler& session_handler,
       SequenceHandler& sequence_handler,
+      const std::string& id,
       const std::string& title) :
-      ScatterPlotWidget(session_handler, sequence_handler, title) {};
+      ScatterPlotWidget(session_handler, sequence_handler, id, title) {};
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/ui/SpectraPlotWidget.h:22]: Good idea differentiating the id from the title.

File: src/smartpeak/include/SmartPeak/ui/Widget.h [hunk 7]
```diff
@@ -76,8 +85,22 @@ namespace SmartPeak
   class GenericTableWidget : public Widget
   {
   public:
-    GenericTableWidget(const std::string& table_id)
-      : table_id_(table_id) {};
+    typedef void(SessionHandler::* DataGeterMethod)(const SequenceHandler& sequence_handler, SessionHandler::GenericTableData& generic_table_data);
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/ui/Widget.h]: I think you mean `DataGetterMethod`?

Comment (dmccloskey) [reply src/smartpeak/include/SmartPeak/ui/Widget.h]: Very cool use of typedef by the way.

File: src/examples/source/SmartPeakGUI.cpp [hunk 8]
```diff
@@ -41,6 +41,9 @@
 #include <SmartPeak/ui/Workflow.h>
 #include <SmartPeak/ui/StatisticsWidget.h>
 #include <SmartPeak/ui/InfoWidget.h>
+#include <SmartPeak/ui/RunWorkflowWidget.h>
```
Comment (dmccloskey) [inline src/examples/source/SmartPeakGUI.cpp:44]: The modularity afforded by the additional widgets is nice to see.

**Review by dmccloskey — COMMENTED** [review]
Very nice WIP!

Are there any particular areas you wanted feedback on?  It is a lot of changes so if there is something specific you wanted an opinion on, please let me know and I can provide more specific feedback.

Also, I think it would be a good idea to have a meeting where you can go over the changes that you did once ready.  You are using a lot of very nice design concepts and techniques that would be informative for myself to understand more your rational behind and I think educational to myself and all in general.

**Comment by bertrandboudaud** [comment]
> Very nice WIP!
> 
> Are there any particular areas you wanted feedback on? It is a lot of changes so if there is something specific you wanted an opinion on, please let me know and I can provide more specific feedback.
> 
> Also, I think it would be a good idea to have a meeting where you can go over the changes that you did once ready. You are using a lot of very nice design concepts and techniques that would be informative for myself to understand more your rational behind and I think educational to myself and all in general.

Thanks for the feedback! indeed the expected feedback was if you feel it's going to the right direction. I will present the modification at the end of the demo meeting next week :)

File: src/smartpeak/include/SmartPeak/ui/Widget.h (hunk 7 above)
Comment (bertrandboudaud) [reply src/smartpeak/include/SmartPeak/ui/Widget.h]: > I think you mean `DataGetterMethod`?

right, it's now fixed

File: src/smartpeak/include/SmartPeak/ui/RunWorkflowWidget.h [hunk 9]
```diff
@@ -35,16 +35,20 @@ namespace SmartPeak
...
-    void setSessionHandler(SessionHandler& session_handler);
-    void setWorkflowManager(WorkflowManager& workflow_manager);
 
   protected:
     bool popup_file_picker_ = false;
     FilePicker file_picker_;
-    ApplicationHandler* application_handler_ = nullptr;
-    SessionHandler* session_handler_ = nullptr;
-    WorkflowManager* workflow_manager_ = nullptr;
+    ApplicationHandler& application_handler_;
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/ui/RunWorkflowWidget.h:50]: Good call on the update from `*` to `&`

File: src/smartpeak/include/SmartPeak/ui/Widget.h [hunk 10]
```diff
@@ -212,8 +220,8 @@ namespace SmartPeak
   {
   public:
     LinePlot2DWidget(const std::string title = "") : GenericGraphicWidget(title)  {};
-    void setValues(const Eigen::Tensor<float, 2>* x_data,
-      const Eigen::Tensor<float, 2>* y_data,
+    void setValues(const Eigen::Tensor<float, 2>& x_data,
+      const Eigen::Tensor<float, 2>& y_data,
       const Eigen::Tensor<std::string, 1>* x_labels,
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/ui/Widget.h:225]: Can the raw pointers here be replaced as well?

File: src/smartpeak/include/SmartPeak/ui/Widget.h [hunk 11]
```diff
@@ -85,14 +85,20 @@ namespace SmartPeak
...
   {
   public:
-    typedef void(SessionHandler::* DataGeterMethod)(const SequenceHandler& sequence_handler, SessionHandler::GenericTableData& generic_table_data);
+    typedef void(SessionHandler::* DataGetterMethod)(const SequenceHandler& sequence_handler, SessionHandler::GenericTableData& generic_table_data);
     typedef Eigen::Tensor<bool, 1>(SessionHandler::* DataFilterMethod)(const Eigen::Tensor<std::string, 2>& to_filter) const;
 
+    /*
+    * GenericTableWidget.
+    * 
+    * A data getter can be provided so that data will be retreive from SessionHandler automatically in the draw method.
```
Comment (dmccloskey) [inline src/smartpeak/include/SmartPeak/ui/Widget.h:94]: In my opinion, this is quite nice how you have maintained the separation of the logic with the data as well as the methods for getting the data and filtering it.  This should make things much easier down the road.

**Review by dmccloskey — APPROVED** [review]
It looks good from my end.  However, it would be great to get other eyes on the code @ahmedskhalil and @ChristopherAbram .  I am assuming you extensively tested the UI after the changes, but it might be useful if @ChristopherAbram or @ahmedskhalil could test out the branch as well if possible.

**Pull request merged**